### PR TITLE
Update `clone!` and `closure!` macro to new syntax

### PR DESCRIPTION
**Pull request by sdroege**
This works correctly with rustfmt now, and thanks to using syn correctly also gives much better user experience with rust-analyzer.

Also the `closure!` macro now supports plain weak references too.

Fixes https://github.com/gtk-rs/gtk-rs-core/issues/1394

### CODE REVIEW

**Comment by sdroege** [comment]
I'll also update the closure macro in a bit once this is cleaned up (and re-use part of the code here I guess).

File: glib-macros/src/lib.rs
Comment (sdroege) [inline glib-macros/src/lib.rs:209]: Are there opinions on this? Should it be called something like `#[default_on_weak_upgrade_failure(panic)]` and `#[default_on_weak_upgrade_failure(return)]` or so instead?

Comment (GuillaumeGomez) [reply glib-macros/src/lib.rs:209]: Simpler is better I think so as is looks good to me.

Comment (pbor) [reply glib-macros/src/lib.rs:209]: Not strong opinion, but I would prefer to have `clone_or!(false, ...` and `clone_or_panic!` variants and do not allow `#[weak]` in the simple `clone!`.

**Comment by sdroege** [comment]
closure macro is also covered now

File: glib-macros/src/clone_old.rs
Comment (sdroege) [inline glib-macros/src/clone_old.rs:858]: `clone_old.rs` and `closure_old.rs` are exactly the same as the previous versions plus the addition of this for generating deprecation warnings.

github unfortunately doesn't detect that they're basically the same files.

With the next major release (0.21) we would drop the old code.

File: glib-macros/src/lib.rs
Comment (sdroege) [reply glib-macros/src/lib.rs:210]: Having separate variants seems like something we'll repeatedly have to tell people about when they ask why `#[weak]` does not work

File: glib-macros/src/clone.rs
Comment (GuillaumeGomez) [inline glib-macros/src/clone.rs:22]: Instead of checking like that, why not implement `TryFrom<str>` (or `TryFromStr`, never remember the name)? Like that you can make `Capture::new` take `CaptureKind` argument.

File: glib-macros/src/clone.rs
Comment (GuillaumeGomez) [inline glib-macros/src/clone.rs:111]: Wouldn't it be simpler to ask for an ident directly?

```rust
#[rename_to = something]
```

doesn't seem that bad either.

File: glib-macros/src/clone.rs
Comment (GuillaumeGomez) [inline glib-macros/src/clone.rs:216]: You forgot to close the paren in your error message. ;)

File: glib-macros/src/clone.rs
Comment (GuillaumeGomez) [inline glib-macros/src/clone.rs:294]: It contradicts your sentence just above which says "none, one or more attributes".

File: glib-macros/src/clone.rs
Comment (GuillaumeGomez) [inline glib-macros/src/clone.rs:326]:
```suggestion
                        &format!("The `default_return` attribute must be followed by 0 or 1 attributes, found {}", attrs.len()),
```

File: glib-macros/src/clone.rs
Comment (GuillaumeGomez) [inline glib-macros/src/clone.rs:336]: Same as above.

File: glib-macros/src/clone.rs
Comment (GuillaumeGomez) [inline glib-macros/src/clone.rs:356]:
```suggestion
                        "Async blocks must be moved. Use the `move` keyword to do so",
```

File: glib-macros/src/clone.rs
Comment (GuillaumeGomez) [inline glib-macros/src/clone.rs:366]: Same as above. You should maybe write a small common function to only write the message in one place (same for the wrong number of attributes above).

File: glib-macros/src/clone.rs
Comment (GuillaumeGomez) [inline glib-macros/src/clone.rs:418]: I think a `match` would look a bit better:

```suggestion
        let default_return_closure = match self.default_return {
             DefaultReturn::Return(Some(ref closure)) => Some(quote! {
                 let #default_return_closure_ident = #closure;
            }),
            _ => None,
        };
```

File: glib-macros/src/closure.rs
Comment (GuillaumeGomez) [inline glib-macros/src/closure.rs:31]: Contradicts above comment.

File: glib-macros/src/closure.rs
Comment (GuillaumeGomez) [inline glib-macros/src/closure.rs:68]: Same as mentioned for the other file.

File: glib-macros/src/closure.rs
Comment (GuillaumeGomez) [inline glib-macros/src/closure.rs:78]: Same as mentioned for the other file.

File: glib-macros/src/lib.rs
Comment (GuillaumeGomez) [inline glib-macros/src/lib.rs:109]:
```suggestion
/// your closure called. In this case, you need to use `#[weak_allow_none]` instead of `#[weak]`:
```

File: glib-macros/src/clone.rs
Comment (sdroege) [reply glib-macros/src/clone.rs:111]: I didn't know that was possible. Better, thanks!

File: glib-macros/src/clone.rs
Comment (sdroege) [reply glib-macros/src/clone.rs:356]: The block is not moved, the captured variables are

File: glib-macros/src/clone.rs
Comment (sdroege) [reply glib-macros/src/clone.rs:366]: Both messages are actually different though

File: glib-macros/src/clone.rs
Comment (GuillaumeGomez) [reply glib-macros/src/clone.rs:356]: Proof that the error message needs improvement. :wink:

**Comment by sdroege** [comment]
@GuillaumeGomez Updated everything, thanks for the review :)

File: glib-macros/src/clone.rs
Comment (GuillaumeGomez) [inline glib-macros/src/clone.rs:342]: It's definitely worth having a common function since it has the same text and the logic as the error above.

File: glib-macros/src/closure.rs
Comment (GuillaumeGomez) [inline glib-macros/src/closure.rs:83]: Same as mentioned above, common function please. :)

File: glib-macros/src/closure_old.rs
Comment (GuillaumeGomez) [inline glib-macros/src/closure_old.rs:136]: You could use (not a blocker, just a style thing, I personally have no preference, just wanted to mention it):

```suggestion
                        "`{name}`: Field accesses are not allowed as is, you must rename it!",
```

File: glib-macros/src/closure_old.rs
Comment (GuillaumeGomez) [inline glib-macros/src/closure_old.rs:168]: Didn't see any test for this error. Please add one.

File: glib-macros/src/closure_old.rs
Comment (GuillaumeGomez) [reply glib-macros/src/closure_old.rs:136]: Please ignore, old file.

File: glib-macros/src/closure.rs
Comment (GuillaumeGomez) [inline glib-macros/src/closure.rs:51]: Didn't see any test for this error. Please add one.

File: glib-macros/src/clone.rs
Comment (GuillaumeGomez) [inline glib-macros/src/clone.rs:34]: Didn't see any test for this error. Please add one.

File: glib-macros/src/clone.rs
Comment (GuillaumeGomez) [inline glib-macros/src/clone.rs:63]: Didn't see any test for this error. Please add one.

File: glib-macros/src/closure.rs
Comment (GuillaumeGomez) [reply glib-macros/src/closure.rs:51]: Also, please put `watch` between backticks.

File: glib-macros/src/closure.rs
Comment (GuillaumeGomez) [inline glib-macros/src/closure.rs:58]: Didn't see any test for this error. Please add one.

File: glib-macros/src/closure.rs
Comment (GuillaumeGomez) [inline glib-macros/src/closure.rs:70]: Didn't see any test for this error. Please add one.

File: glib-macros/src/clone.rs
Comment (sdroege) [reply glib-macros/src/clone.rs:342]: Except that it's not :rofl: It's looking at different arrays, different indices, does the length handling different.

File: glib-macros/src/clone.rs
Comment (sdroege) [reply glib-macros/src/clone.rs:366]: My message says exactly what is required: people have to add the `move` keyword.

Do you have a suggestion for a better message?

File: glib-macros/src/clone.rs
Comment (GuillaumeGomez) [reply glib-macros/src/clone.rs:342]: I mean for the error, you can pass value, slice and everything.

File: glib-macros/src/lib.rs
Comment (DaKnig) [reply glib-macros/src/lib.rs:210]: how about `#[on_upgrade_failure(return)]` ? `default` doesnt mean anything in this context...

File: glib-macros/src/clone.rs
Comment (DaKnig) [inline glib-macros/src/clone.rs:230]: let ... else

File: glib-macros/src/lib.rs
Comment (sdroege) [reply glib-macros/src/lib.rs:210]: I'd prefer something shorter. We'll have some time for bikeshedding here before this is merged :)

File: glib-macros/src/closure.rs
Comment (sdroege) [reply glib-macros/src/closure.rs]: There are no tests for `closure!` error messages at all

Comment (sdroege) [reply glib-macros/src/closure.rs]: I.e. I'll add those

**Review by GuillaumeGomez — APPROVED** [review]

File: glib-macros/src/lib.rs
Comment (sdroege) [reply glib-macros/src/lib.rs:210]: I prefer the variant you list under `my preference`. Also over what is implemented here right now. Unless there are more opinions, I'll change it to that and implement the two other variants you mentioned.

Comment (sdroege) [reply glib-macros/src/lib.rs:210]: As discussed on Matrix that might be too implicit and cause `false` to be returned unintentionally.

Comment (sdroege) [reply glib-macros/src/lib.rs:210]: `upgrade_or_panic` seems most descriptive to me

File: glib-macros/src/clone.rs
Comment (DaKnig) [inline glib-macros/src/clone.rs:194]:
```suggestion
                let #alias = std::borrow::ToOwned::to_owned(&*#name);
```

Comment (sdroege) [reply glib-macros/src/clone.rs:194]: Why?

File: glib-macros/src/lib.rs
Comment (sdroege) [reply glib-macros/src/lib.rs:210]: So let's summarize: `#[upgrade_or_else]` for the closure, `#[upgrade_or]` for a plain value (needs to be `Copy`!), `#[upgrade_or_default]` for returning `Default::default()`, `#[upgrade_or_panic]` for panicking and if nothing is specified then `()` is returned (and it will fail to compile if the closure is supposed to return a different type).

**Comment by sdroege** [comment]
Updated everything accordingly and will merge once the CI is happy. Everything else can happen on top of this

**Pull request merged**
